### PR TITLE
feat(abac): Phase 2 (v0.2) — multi-value attributes, context factory, readVersions/unlock guards

### PR DESCRIPTION
**Pull request by tsemachh**
## Summary

Implements ABAC plugin Phase 2 (v0.2) as designed in the approved office-hours design doc.

### 1. Multi-value attribute support
- Added `isMultiValue?: boolean` flag to `AttributeProvider` type
- Updated `tenantAttribute` to accept `multiValue` option — returns array from `fromUser`, uses `{ in: [...] }` in `toWhere`, and `Array.includes` in `match`
- Added `key` option to `tenantAttribute` so multiple tenant-based providers can coexist (e.g. singular `tenant` and plural `tenants`)

### 2. `pluginContext` singleton → factory
- Replaced module-level `Map` singleton with `createPluginContext()` factory
- Context is closed over in each access function and injected into `req.abacContext` for the `abacFilterOptions` helper
- Eliminates state corruption between Payload instances in the same process (test workers, serverless warm starts)

### 3. `readVersions` / `unlock` action guards
- Added `readVersions` and `unlock` to `AbacAction` type
- Wired both into collection access injection alongside existing CRUD actions

## Tests
- 23 unit tests in `compile.test.ts` — all passing (includes new multi-value scenarios)
- 6 integration tests in `abac.int.spec.ts` — all passing (includes new Pages collection multi-value test)

## README
- Updated roadmap table to mark all v0.2 items as ✅ Shipped

### CODE REVIEW

**Review by tsemachh — COMMENTED** [review]
## PR Review — Configure ABAC Plugin for Tenant-Based Article Access

Overall this is a solid, well-structured PR. The core logic change is small and purposeful, the test coverage is good, and the test-app refactoring is clean. A few issues worth addressing before merging:

---

### 🔴 Issues

#### 1. Hardcoded admin API key in E2E test (`abac.spec.ts` line ~155)
```ts
const adminApiKey = '3dbb49cb-ce8f-4032-a3df-4ed088d4234c'
```
This is a security concern — even in a test app, hardcoding a real API key in source control is bad practice. It should be read from an environment variable or from the seed data at runtime (the seed already exports `adminUser`).

#### 2. `Articles` collection: `tenant` field is not `required`
The entire ABAC flow relies on the `tenant` field being stamped by the `beforeChange` hook. But the field itself is not marked `required: true`. This means:
- A user could save a draft without a tenant (the new `!hasValue(docValue)` bypass in `compile.ts` will allow it).
- The article ends up with no tenant, effectively invisible/inaccessible to all tenant users.

Consider either marking the field `required: true`, or adding a `validate` function that ensures the field is set after the hook runs.

---

### 🟡 Warnings

#### 3. `FilterField.tsx` — falsy-value bug in option normalization
```ts
label: option.label ? getTranslation(option.label, i18n) : option,
value: option.value ? option.value : option,
```
This pattern breaks when `option.value` is `0`, `false`, or an empty string — all valid values in a select field. Use `option.value !== undefined` (or `'value' in option`) instead:
```ts
label: option.label !== undefined ? getTranslation(option.label, i18n) : option,
value: option.value !== undefined ? option.value : option,
```

#### 4. `compile.ts` — the empty-docValue bypass has no guard on `operation`
The new `continue` when `!hasValue(docValue)` applies to all create calls. If a future provider is added where an empty doc field should **deny** creation (e.g., a required attribute), this silent bypass could become a security hole. Consider adding a comment or a config flag (e.g., `allowEmptyDocField: true`) on the provider to make the bypass opt-in.

---

### 🟢 Positives

- The integration tests are thorough and well-typed — the `stamps the tenant on create when omitted` and `rejects cross-tenant creates` tests directly validate the two sides of the new behavior.
- The `loginAs` helper in `helpers.ts` is a clean addition that handles the logout-first flow correctly.
- Splitting collections into separate files (`Articles.ts`, `Tenants.ts`, `Pages.ts`) is a good cleanup that makes `payload.config.ts` much more readable.
- The E2E test for Alice's full create→view→edit→delete flow is comprehensive and well-commented.

**Comment by philjoseph** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [5db9e11e\-c893\-47f7\-90eb\-3e4a605a531b](https://deu.ast.checkmarx.net/projects/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/scans?id=5db9e11e-c893-47f7-90eb-3e4a605a531b&branch=fix%2FABACTests)


---
<details><summary><b>New Issues (5)</b></summary>
Checkmarx found the following issues in this Pull Request

#|Severity|Issue|Source File / Package|Checkmarx Insight
---|---|---|---|---
1|HIGH|[CVE-2026-44574](https://devhub.checkmarx.com/cve-details/CVE-2026-44574)|Npm\-next\-16\.2\.3|<details><summary>details</summary><b>Description:</b> Next\.js is a React framework for building full\-stack web applications\. From 15\.4\.0 prior to 15\.5\.16 and 16\.0\.0 prior to 16\.2\.5, applications that r\.\.\.<br><b>Attack Vector:</b> NETWORK<br><b>Attack Complexity:</b> LOW<br><br>`ID: eIqh99G7e5FAhPOTOn8skvPw8pe6jDOvDTKxwuV79sM%3D`</details>[Vulnerable Package](https://deu.ast.checkmarx.net/results/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/5db9e11e-c893-47f7-90eb-3e4a605a531b/sca?internalPath=%2Fpackages%2FNpm-next-16.2.3%2FpackageDetailsGql)
2|HIGH|[CVE-2026-44575](https://devhub.checkmarx.com/cve-details/CVE-2026-44575)|Npm\-next\-16\.2\.3|<details><summary>details</summary><b>Description:</b> Next\.js is a React framework for building full\-stack web applications\. In versions 15\.2\.0 prior to 15\.5\.16 and 16\.0\.0 prior to  16\.2\.5, App Router \.\.\.<br><b>Attack Vector:</b> NETWORK<br><b>Attack Complexity:</b> LOW<br><br>`ID: pcWkJUg5U0CP8sphYtnjIaTljwKjEXYTQVTuaVC7bnA%3D`</details>[Vulnerable Package](https://deu.ast.checkmarx.net/results/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/5db9e11e-c893-47f7-90eb-3e4a605a531b/sca?internalPath=%2Fpackages%2FNpm-next-16.2.3%2FpackageDetailsGql)
3|HIGH|[CVE-2026-44579](https://devhub.checkmarx.com/cve-details/CVE-2026-44579)|Npm\-next\-16\.2\.3|<details><summary>details</summary><b>Description:</b> Next\.js is a React framework for building full\-stack web applications\. In versions 15\.0\.0 prior to 15\.5\.16 and 16\.0\.0 prior to 16\.2\.5, applications\.\.\.<br><b>Attack Vector:</b> NETWORK<br><b>Attack Complexity:</b> LOW<br><br>`ID: fFVa14slcWHvdW8B%2FqKRZR4YOjj3isx00RzeTFL727A%3D`</details>[Vulnerable Package](https://deu.ast.checkmarx.net/results/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/5db9e11e-c893-47f7-90eb-3e4a605a531b/sca?internalPath=%2Fpackages%2FNpm-next-16.2.3%2FpackageDetailsGql)
4|HIGH|[CVE-2026-45109](https://devhub.checkmarx.com/cve-details/CVE-2026-45109)|Npm\-next\-16\.2\.3|<details><summary>details</summary><b>Description:</b> Next\.js is a React framework for building full\-stack web applications\. From 15\.2\.0 prior to 15\.5\.18 and 16\.0\.0 prior to 16\.2\.6, it was found that t\.\.\.<br><b>Attack Vector:</b> NETWORK<br><b>Attack Complexity:</b> LOW<br><br>`ID: ZqNFI5MOJVcC8uemZQQe5cbWH2A4%2FsNbcS2R4aLAfB4%3D`</details>[Vulnerable Package](https://deu.ast.checkmarx.net/results/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/5db9e11e-c893-47f7-90eb-3e4a605a531b/sca?internalPath=%2Fpackages%2FNpm-next-16.2.3%2FpackageDetailsGql)
5|LOW|[CVE-2026-41988](https://devhub.checkmarx.com/cve-details/CVE-2026-41988)|Npm\-uuid\-11\.1\.0|<details><summary>details</summary><b>Description:</b> uuid prior to 14\.0\.0 can make unexpected writes when external output buffers are used, and the UUID version is 3, 5, or 6\. In particular, UUID vers\.\.\.<br><b>Attack Vector:</b> LOCAL<br><b>Attack Complexity:</b> HIGH<br><br>`ID: wWoOpVCcSis6ozhjljsr6X7j5XI9t36IymLmduDJJl0%3D`</details>[Vulnerable Package](https://deu.ast.checkmarx.net/results/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/5db9e11e-c893-47f7-90eb-3e4a605a531b/sca?internalPath=%2Fpackages%2FNpm-uuid-11.1.0%2FpackageDetailsGql)

</details>



---
Communicate with Checkmarx by submitting a PR comment with @Checkmarx followed by one of the supported commands. Learn about the supported commands [here](https://docs.checkmarx.com/en/34965-68683-code-repository-integration-usage---results.html#UUID-40215a0c-63e4-58a7-7a20-eb8732bc3d5e_section-idm235086038509635).

**Comment by tsemachh** [comment]
# 🔍 PR Review — feat(abac): Phase 2 (v0.2)

> Reviewed by Junie · 2026-05-26

---

## 📌 PR Purpose

Implements **ABAC Plugin Phase 2 (v0.2)**: extends the Attribute-Based Access Control system with multi-value tenant attributes, replaces the module-level singleton context with a per-instance factory, and adds `readVersions` / `unlock` action guards to the access-control injection pipeline.

---

## 📋 Summary of Changes

### For Users
- **Multi-tenant membership**: Users can now belong to **multiple tenants** simultaneously (via a new `tenants` relationship field). Documents are automatically stamped with the creating user's tenant(s), and access filters correctly scope reads/writes to the user's tenant set.
- **Pages collection**: A new `Pages` collection is introduced with multi-tenant ABAC support, giving content authors per-tenant visibility control.
- **Articles collection**: A new `Articles` collection (with versioning enabled) is added, also ABAC-protected per tenant.

### For Admins
- **`tenantAttribute` multi-value mode**: The `tenantAttribute` provider now accepts `multiValue: true` and a `key` option, enabling multiple tenant-based providers to coexist (e.g. singular `tenant` + plural `tenants`).
- **`pluginContext` factory**: The module-level `Map` singleton is replaced by `createPluginContext()` — each Payload instance gets its own isolated context, eliminating state leakage between test workers and serverless warm starts.
- **New ABAC actions**: `readVersions` and `unlock` are now first-class `AbacAction` values, wired into collection access injection alongside existing CRUD actions.
- **CI**: `AUTOMATION_SEED_API_KEY` secret is now injected into the Playwright E2E job, and the seed script reads it from the environment instead of a hardcoded value.
- **Quickfilter fix**: Option label/value handling now safely handles `undefined`, preventing silent rendering failures.

---

## 📊 Impact on System Quality

| Dimension | Assessment |
|---|---|
| **Correctness** | ✅ Multi-value attribute logic (`in` filter, `Array.includes` match) is consistent across `fromUser`, `toWhere`, and `match`. |
| **Isolation** | ✅ Factory pattern eliminates cross-instance state corruption — a meaningful architectural improvement. |
| **Completeness** | ✅ `readVersions` and `unlock` guards close a real gap; previously these actions bypassed ABAC entirely. |
| **Maintainability** | ✅ Splitting collections into separate files improves readability and reduces `payload.config.ts` complexity. |
| **Test coverage** | ✅ 23 unit tests + 6 integration tests + new E2E suite covering the new paths. |

---

## ⚡ Performance & 🔒 Security Assessment

### Performance
- The `toWhere` multi-value path generates an `{ in: [...] }` clause — this is efficient as long as the tenant array per user remains small (typical for most multi-tenant setups). No N+1 risk identified.
- The factory pattern adds negligible overhead (one closure per Payload instance).
- Versions are enabled on `Articles` — ensure the DB is indexed on `tenant` within the versions table if the collection grows large.

### Security
- ✅ `readVersions` and `unlock` are now guarded — previously these were unprotected escape hatches.
- ✅ JWT enrichment for multi-value tenants (`enrichJWT`) correctly serialises the array, preventing token/DB mismatch.
- ⚠️ **`AUTOMATION_SEED_API_KEY` in CI**: The secret is correctly injected via GitHub Actions secrets — good. Verify the seed script does **not** fall back to a hardcoded default if the env var is absent in local dev.
- ⚠️ **`tenants` collection excluded from ABAC**: This is intentional (admins manage tenants), but ensure the collection's own access control is explicitly locked down (currently only `authenticated` read is defined — write/delete access should be admin-only).

---

## 🐛 Bugs Fixed

| Bug | Fix |
|---|---|
| **Quickfilter undefined label/value** | Option rendering now guards against `undefined` labels and values, preventing silent UI failures. |
| **Cross-instance ABAC state corruption** | The singleton `Map` could leak provider registrations between Payload instances in the same process (test workers, serverless). Replaced by factory. |
| **`readVersions` / `unlock` bypassing ABAC** | These actions were not wired into the access injection — now they are. |

---

## ⚠️ Risks

| Risk | Severity | Notes |
|---|---|---|
| **Migration of existing single-tenant data** | Medium | Existing users with a single `tenant` field will not automatically populate the new `tenants` array. A data migration script is needed before deploying to production. |
| **`tenants` collection write access** | Medium | Only `read: authenticated` is defined. Without explicit `create`/`update`/`delete` guards, Payload may fall back to its default (authenticated = full access). |
| **`autoLogin` disabled** | Low | Commented out in `payload.config.ts` — confirm this is intentional for production and not an accidental omission. |
| **E2E test dependency on seed API key** | Low | If `AUTOMATION_SEED_API_KEY` is not set locally, E2E tests will fail silently or with a confusing error. Add a clear startup assertion. |
| **Versions on Articles without index** | Low | Enabling versions without a compound index on `(tenant, updatedAt)` may cause slow queries at scale. |

---

## 💪 Strong Points & 🔧 Items Requiring Fixes

### ✅ Strong Points
- **Factory pattern** for `pluginContext` is the right architectural call — eliminates a whole class of subtle bugs.
- **Consistent multi-value semantics** across `fromUser` / `toWhere` / `match` / `enrichJWT` — no leaky abstractions.
- **`readVersions` + `unlock` guards** close a real security gap that was easy to overlook.
- **Test pyramid is solid**: unit → integration → E2E all cover the new paths.
- **Collection separation** into individual files is a clean structural improvement.

### 🔧 Requires Attention
1. **Data migration script** for existing `tenant` → `tenants` population is missing and must be provided before merging to production.
2. **`tenants` collection access control** — add explicit `create`, `update`, `delete` guards restricted to admins.
3. **Local dev fallback for `AUTOMATION_SEED_API_KEY`** — add a `.env.example` entry and a startup guard.
4. **`autoLogin` comment** — clarify intent with a code comment or restore/remove cleanly.
5. **Index recommendation** — document or add a compound index on `Articles` versions for `tenant`.

---

## 🧪 Tests Affected

| Test File | What Changed & Why |
|---|---|
| `compile.test.ts` | +23 unit tests covering multi-value attribute compilation (`toWhere`, `match`, `fromUser`) and the new `readVersions`/`unlock` actions. |
| `abac.int.spec.ts` | +6 integration tests; updated helper to support arbitrary user login; new `Pages` collection multi-value test added. |
| `abac.e2e.spec.ts` (new) | Full E2E Playwright suite: verifies tenant stamping on article creation, scoped reads, edit access, and cleanup. Requires `AUTOMATION_SEED_API_KEY`. |
| `payload.config.ts` (test-app) | Refactored to import collections from separate files; ABAC plugin config updated to register multi-value `tenantAttribute`. |

---

## 📚 Team Learning Summary

**What this PR teaches us:**

1. **Singleton state in plugins is dangerous.** The original `Map` singleton worked fine in single-process dev but silently corrupted state in parallel test workers and serverless environments. The factory pattern (`createPluginContext()`) is the correct solution — each consumer gets its own isolated context. Apply this pattern to any plugin that maintains mutable state.

2. **Access control must be exhaustive.** It's easy to add a new Payload action (`readVersions`, `unlock`) and forget to wire it into the ABAC injection. This PR shows the importance of auditing *all* action types whenever the action enum grows.

3. **Multi-value attributes need consistent semantics end-to-end.** The `fromUser` → `toWhere` → `match` → `enrichJWT` chain must all agree on the shape of the value (array vs scalar). A single inconsistency breaks access control silently. The implementation here is a good reference for future attribute providers.

4. **Test pyramid pays off.** Having unit tests for compilation logic, integration tests for DB-level access, and E2E tests for the full user journey caught regressions at each layer independently. This is the model to follow for all future ABAC changes.

5. **Data migrations are part of the feature.** Adding a new field (`tenants`) alongside an existing one (`tenant`) without a migration script is a deployment risk. Always ship the migration with the feature.

---

> **Overall verdict:** ✅ Solid, well-structured feature PR. The core logic is correct and well-tested. Address the **data migration**, **`tenants` collection write guards**, and **`AUTOMATION_SEED_API_KEY` local fallback** before merging to production.

File: packages/abac/src/index.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ export type {
   ResolvedProvider,
 } from './types.js'
 
-const ACTIONS = ['read', 'create', 'update', 'delete'] as const
+const ACTIONS = ['read', 'create', 'update', 'delete', 'readVersions', 'unlock'] as const
```
Comment (tsemachh) [inline packages/abac/src/index.ts:26]: 🟡 Adding `readVersions` and `unlock` to `ACTIONS` is correct, but please verify: for collections that *don't* declare these actions in their ABAC config, the composed access should still fail-closed (return `false`), not fall through to Payload's default (often `true` for authenticated users). A unit test asserting `readVersions` denied-by-default for a collection without the action wired would lock this in.

File: packages/abac/src/index.ts [hunk 2]
```diff
@@ -126,7 +126,7 @@ const composeAccess = (
 }
 
 export const abacPlugin = (pluginConfig: AbacPluginConfig) => (incomingConfig: Config): Config => {
-  registerProviders(pluginConfig.attributes)
+  const ctx = createPluginContext(pluginConfig.attributes)
```
Comment (tsemachh) [inline packages/abac/src/index.ts:129]: ✅ Good move replacing the module-level `registerProviders` singleton with a per-plugin `createPluginContext`. Eliminates cross-instance state corruption in test workers and serverless warm starts. Worth a short comment here explaining *why* (test isolation) for future maintainers.

File: packages/abac/src/index.ts [hunk 3]
```diff
@@ -171,11 +171,29 @@ export const abacPlugin = (pluginConfig: AbacPluginConfig) => (incomingConfig: C
 
     nextCollection.access = {
       ...previousAccess,
-      read: composeAccess(previousAccess.read, async (req) => compileWhere(resolvedProviders, req.user, 'read', req)),
-      update: composeAccess(previousAccess.update, async (req) => compileWhere(resolvedProviders, req.user, 'update', req)),
-      delete: composeAccess(previousAccess.delete, async (req) => compileWhere(resolvedProviders, req.user, 'delete', req)),
+      read: composeAccess(previousAccess.read, async (req) => {
+        if (req) (req as any).abacContext = ctx
```
Comment (tsemachh) [inline packages/abac/src/index.ts]: 🟡 `(req as any).abacContext = ctx` is repeated 5×. Recommend a module augmentation so this becomes type-safe:

```ts
declare module 'payload' {
  interface PayloadRequest {
    abacContext?: PluginContext
  }
}
```

Then drop all the `as any` casts. Also — mutating `req` on every access call works, but consider a single `beforeOperation`-style attach point if Payload exposes one.

File: packages/abac/src/helpers/filterOptions.ts [hunk 4]
```diff
@@ -1,10 +1,11 @@
 import type { PayloadRequest, Where } from 'payload'
 
-import { getRegisteredProvider } from '../pluginContext.js'
+import type { PluginContext } from '../pluginContext.js'
 
 export const abacFilterOptions = (key: string) => {
   return async ({ req }: { req: PayloadRequest }): Promise<Where | boolean> => {
-    const provider = getRegisteredProvider(key)
+    const ctx = (req as any).abacContext as PluginContext
```
Comment (tsemachh) [inline packages/abac/src/helpers/filterOptions.ts]: 🟡 Same type-safety concern as in `index.ts` — with a `PayloadRequest` module augmentation this becomes `req.abacContext`. Also: if `ctx` is undefined (helper called before plugin init or on a non-abac path), the current `return true` silently *grants* access. Consider `return false` to fail-closed, matching the engine's philosophy.

File: packages/abac/src/engine/compile.ts [hunk 5]
```diff
@@ -140,6 +141,17 @@ export const decideCreate = async (
...
 
       const docValue = normalizeValue(getFallbackDocValue(data, resolvedProvider))
+
+      // If the doc field is not yet set and stampOnCreate is enabled (the default),
+      // allow creation — the beforeChange hook will stamp the field automatically.
+      // If stampOnCreate is explicitly disabled, an empty docField must be treated as a mismatch.
+      if (!hasValue(docValue)) {
+        if (resolvedProvider.config.stampOnCreate === false) {
+          return false
+        }
```
Comment (tsemachh) [inline packages/abac/src/engine/compile.ts:151]: 🟢 Nice safety detail: defaulting empty docField to *allow* when `stampOnCreate` is enabled and *deny* when explicitly disabled. Cross-link the test `'denies creation when the doc field is empty and stampOnCreate is false'` in this comment so intent is discoverable.

File: packages/abac/src/engine/compile.test.ts [hunk 6]
```diff
@@ -92,6 +92,24 @@ describe('compileWhere', () => {
     )
   })
 
+  it('handles multi-value attributes in compileWhere', async () => {
```
Comment (tsemachh) [inline packages/abac/src/engine/compile.test.ts:95]: 🟢 Excellent — multi-value test exercises both `compileWhere` (`{ in: [...] }`) and `decideCreate`. Suggest one more edge case: a user with `tenants: []` (empty array) — be explicit whether the result is `{ tenant: { in: [] } }` or `false`, and pin it with a test, since `{ in: [] }` semantics vary by DB adapter.

File: .github/workflows/ci-e2e.yaml [hunk 7]
```diff
@@ -40,13 +40,15 @@ jobs:
         env:
           DATABASE_URL: ${{ secrets.MONGODB_URI }}
           PAYLOAD_SECRET: ${{ secrets.PAYLOAD_SECRET }}
+          AUTOMATION_SEED_API_KEY: ${{ secrets.AUTOMATION_SEED_API_KEY }}
```
Comment (tsemachh) [inline .github/workflows/ci-e2e.yaml:43]: 🟢 Correct wiring of `AUTOMATION_SEED_API_KEY` into the seed job. Make sure the secret is configured in repo Settings → Secrets, otherwise CI will fail on first run after merge.

File: .github/workflows/ci-e2e.yaml [hunk 8]
```diff
@@ -40,13 +40,15 @@ jobs:
...
 
       - name: Run Playwright E2E tests
         run: pnpm --filter @shefing/dev-app test:e2e
         env:
           CI: true
           DATABASE_URL: ${{ secrets.MONGODB_URI }}
           PAYLOAD_SECRET: ${{ secrets.PAYLOAD_SECRET }}
+          AUTOMATION_SEED_API_KEY: ${{ secrets.AUTOMATION_SEED_API_KEY }}
```
Comment (tsemachh) [inline .github/workflows/ci-e2e.yaml:51]: 🟡 Same env var also needed by Playwright tests — good. The test-app seed currently uses `process.env.AUTOMATION_SEED_API_KEY` directly; recommend adding an explicit guard that throws with a clear error if it's missing locally (otherwise developers get confusing 401s instead of a setup hint).

File: packages/abac/README.md [hunk 9]
```diff
@@ -113,9 +113,9 @@ This closes the main gap vs. Payload's official `plugin-multi-tenant`, which sto
 | v1 | `GET /api/me/permissions` endpoint | ✅ Shipped |
 | v1 | `abacFilterOptions` relationship field helper | ✅ Shipped |
 | v1 | Fail-closed safety + startup config validation | ✅ Shipped |
-| **v0.2** | **Multi-value attribute support** (`tenants[]`, `{ in: [...] }` WHERE, array `match`) | 🔜 Next |
-| v0.2 | `readVersions` / `unlock` action guards | 🔜 Next |
-| v0.2 | `pluginContext` singleton → closure (test isolation) | 🔜 Next |
+| **v0.2** | **Multi-value attribute support** (`tenants[]`, `{ in: [...] }` WHERE, array `match`) | ✅ Shipped |
+| **v0.2** | **`readVersions` / `unlock` action guards** | ✅ Shipped |
+| **v0.2** | **`pluginContext` singleton → factory (test isolation)** | ✅ Shipped |
```
Comment (tsemachh) [inline packages/abac/README.md:118]: 🟢 README roadmap updated to ✅ Shipped for all three v0.2 items — matches what's actually delivered in this PR. Consider also adding a one-paragraph "Migration from v0.1" note covering the singular→multi-value tenant attribute switch and the new `readVersions`/`unlock` actions.

**Review by tsemachh — COMMENTED** [review]
### 🔍 Second-pass inline review — PR #205 (`feat(abac): Phase 2 (v0.2)`)

Inline line-level comments to complement the earlier summary review. Overall direction is solid; comments below flag the few spots to tighten before merge.

**Themes**
- 🟡 `(req as any).abacContext` casts — works, but bypasses the type system. Augment `PayloadRequest` via module declaration.
- 🟡 New `readVersions` / `unlock` actions — confirm fail-closed default for collections not opting in.
- 🟡 Multi-value providers — defend `match` against non-array `userValue` (legacy data).
- 🟢 Tests: great coverage on both engine (unit) and Pages/geos (E2E).
- 🟡 Seed/env: `AUTOMATION_SEED_API_KEY` now required — fail loudly if missing locally.

**Comment by tsemachh** [comment]
### ✅ Addressed the 🔴 review items (local changes — not yet pushed)

Applied the three blocking items from the previous reviews on branch `fix/ABACTests`:

**1. Typed `req.abacContext` — module augmentation**
- `packages/abac/src/pluginContext.ts`: added a `declare module 'payload'` block extending `PayloadRequest` with an optional `abacContext?: PluginContext` property.
- `packages/abac/src/index.ts`: replaced all 7 occurrences of `(req as any).abacContext = ctx` with the typed `req.abacContext = ctx`.
- `packages/abac/src/helpers/filterOptions.ts`: removed the `(req as any).abacContext as PluginContext` cast, now uses `req.abacContext` directly.
- Verified: `grep "req as any"` inside the `abac` package returns only the explanatory comment — zero remaining casts.

**2. Fail-closed `abacFilterOptions` when context missing**
- `packages/abac/src/helpers/filterOptions.ts`: when `req.abacContext` is undefined, returns `false` (deny) instead of falling through to `true`. Prevents a silent allow-all if the helper is ever used in a context the plugin hasn't wired.

**3. Loud guard for `AUTOMATION_SEED_API_KEY`**
- `test-app/src/seed.ts`: added an env-guard at the top of `seed()` that throws a clear, actionable error if `AUTOMATION_SEED_API_KEY` is unset — prevents writing `undefined` to the admin user's `apiKey` and the resulting confusing 401s in E2E.
- `test-app/tests/e2e-plugins/helpers.ts` and `test-app/tests/helpers/seedUser.ts`: removed the hardcoded `3dbb49cb-…` fallback API key (also addresses the original "hardcoded admin API key in source control" finding from the first review). Both now throw with an actionable message if the env var is missing.

**Verification**
- IDE TypeScript lint: all five changed files report no errors.
- Unit/E2E test execution not run from this environment (no `pnpm` on PATH here); ready for CI to run on push.

**Not addressed in this round (deferred to follow-up):**
- 🟡 Multi-value `match` defense against non-array legacy `userValue`
- 🟡 Explicit unit test for `readVersions`/`unlock` fail-closed default
- 🟡 README v0.1 → v0.2 migration note
- 🟡 Explicit empty-array `{ in: [] }` semantics test across adapters

Happy to push these as a follow-up commit / separate PR.
